### ISSUE TITLE
MSRV-dependent dependency version resolution

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/cargo#9930]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/cargo/issues/9930)*

Based on [RFC 2495](https://github.com/rust-lang/rfcs/pull/2495) Rust now supports the [`rust-version`](https://doc.rust-lang.org/nightly/cargo/reference/manifest.html#the-rust-version-field) field, which currently works as:

>If the currently selected version of the Rust compiler is older than the stated version, cargo will exit with an error, telling the user what version is require

This behavior is not ideal. If a project depends on crate `foo v0.1.0` with MSRV 1.56, then releasing crate `foo v0.1.1` with MSRV 1.60 will break build of the project on older toolchains after simple `cargo update`. Ideally Cargo would select `foo v0.1.0` on older tolchains, thus preventing the breakage.

RFC 2495 has [described](https://github.com/rust-lang/rfcs/blob/master/text/2495-min-rust-version.md#influencing-version-resolution) MSRV-dependent version resolution as a potential future extension. This issue is intended for design discussions and tracking implementation progress of this feature.

Third-party support
- dependabot/dependabot-core#5423
- renovatebot/renovate#26060

---

This has been approved as [RFC #3537](https://rust-lang.github.io/rfcs/3537-msrv-resolver.html).

Implementation by stabilization milestone
- [x] Improve the `cargo install` error (#12798)
- [x] Improve documentation around `rust-version` (#13056)
- [x] cargo-add (#13608)
  - [x] `cargo add` support (#10653) 
  - [x] Infer users rust-version from `rustc -V` (#13516)
- [x] Attempt to report the MSRV error top-down, rather than bottom up (#13514)
- [x] `cargo update` keeps user informed when using old versions (#13372)
- [x] #13539
- [x] Resolver (#14639, #14754)
  - [x] Prototype (#12560)
  - [x] Switch resolver to "prefer" rather than "enforce" (#12950)
  - [x] De-prioritize dep versions without `rust-version` (#13066)
  - [x] Infer users rust-version from `rustc -V` (#13743)
  - [x] Existing `--ignore-rust-version` support (#13738)
  - [x] `--ignore-rust-version` support for  `cargo update` / `cargo-generate-lockfile` (#13742)
  - [x] #13540
  - [x] Add resolver v3 (#13776)
  - [x] Verify resolver 3 doesn't change cargo-install (#13769, #13790)
  - [x] Verify `cargo update --precise` can select incompatible versions (#13769)
  - [x] Enable resolver v3 with Edition 2024 (#13785)
  - [x] #14569
  - [x] #14724
- [x] Have lockfile format version respect `--ignore-rust-version` (see #12861)
- [ ] `--update-rust-version` support
  - `cargo add`
  - `cargo update`
  - `cargo generate-lockfile`
- [ ] Turn MSRV error into a lint (blocked on #12235) 
- [ ] "auto": Add field with cargo-publish behavior (see also #11451)
- [ ] Update `cargo new` template

Changes from RFC

Unresolved questions
- `cargo add` and the resolver have different behavior for unset `rust-version`, see #13791

Deferred
- Determination for what `cargo install` should do, see also #10903

**Comment by Eh2406** [comment]
Transferred to Cargo.

IMO one of the big problems blocking this is the error messages. The current Resolver has pretty bad messages, it just reports on the last thing that did not work. So In effect the Resolver will just ignore all versions with a different MSRV leaving no record of why newer versions were not considered. This may be on the list of things to consider when/if we use [PubGrub](https://github.com/pubgrub-rs/pubgrub). But that is just my opinion.

**Comment by leo60228** [comment]
Is this actually a good idea? I feel like `rust-version` should solely be a safeguard, and not actually affect behavior. IMO pinning a Rust version without pinning dependencies is user error, and it's better for this to be a hard error instead of doing what might be the wrong thing.

**Comment by newpavlov** [comment]
>Is this actually a good idea?

It was my main motivation for writing the MSRV RFC. Without it I (and many others) consider bumping MSRV a breaking change, since otherwise we risk breaking people's builds on older toolchains (e.g. think of Rust packaged with Debian). Printing an error is just slightly improves ergonomics, no solves the problem at the root.

**Comment by newpavlov** [comment]
@Eh2406
>IMO one of the big problems blocking this is the error messages.

Wouldn't it be viable to implement an MVP with bad error messages first and work on improving them later?

>So In effect the Resolver will just ignore all versions with a different MSRV leaving no record of why newer versions were not considered.

I think it should be fine to start with a simple warning if at least one dependency version was ignored due to the MSRV filtration.

**Comment by MG-96** [comment]
I came across a related problem, where a CI/CD pipeline failed due to a updated dependency which demanded a higher MSRV than specified for the downstream project.

While I am not sure, that doing a automated MSRV-dependent dependency version resolution is a good idea in general, I would like a feature which notifies you if dependencies can't be build with the specified MSRV.
As far as I can tell, it should be no big deal to implement a warning which, when building, notifies you about this.
This may not prevent the problem that a upstream crate may change the MSRV in a semver compatible release in the future, but ensures that there isn't already a unwanted mismatch at development time.

Furthermore I am not completely sure how its meant to be specified: Should the developer ensure, that the MSRV of his Crate is at least as high as all upstream dependencies?

[RFC 2495](https://rust-lang.github.io/rfcs/2495-min-rust-version.html) mentions this, but there seems to be no check implemented right now.

> rust field value will be checked as well. During crate build cargo will check if all upstream dependencies can be built with the specified MSRV. (i.e. it will check if there is exists solution for given crates and Rust versions constraints) Yanked crates will be ignored in this process.

**Comment by MG-96** [comment]
> Cargo from Rust 1.56 and newer should yield an MSRV-specific error when trying to build a dependency that has its rust-version set to something lower than the currently executing version of the compiler.

Just tried to provoke this behaviour with two freshly created crates, and can't really confirm this.
As long as the `edition` and `rust-version` key match, there seems to be no warning or error at all.
Or did I get something wrong here.

When building crate A which depends on crate B, I can set `rust-version` and `edition` to whatever I like for both crates and won't get a error (e.g. rust-version 1.42 for crate A and 1.61 for crate B).
`cargo --version` gives me `cargo 1.61.0 (a028ae4 2022-04-29)`.

**Comment by djc** [comment]
I didn't say you would get an error in that situation. You're building with Cargo 1.61 and both of the crates have 1.61 or older, so there is no error. We currently don't check that depending crates have a newer or equal MSRV compared to the crate they're depending on.

**Comment by tarcieri** [comment]
Perhaps there could be an unstable option similar to `cargo update -Z minimal-versions` to perform a resolution that considers MSRV?

**Comment by epage** [comment]
Highlights of discussion so far:
---

Need for notifying users

>  if this behavior doesn't come with a warning, it's trivially possible that downstream crates no longer get updated to upstream dependencies that fixed security vulnerabilities, for example. While I usually deploy cargo-deny in CI to make sure I'm aware of this, as long as something similar is not integrated in Cargo it would be a little scary that semver-compatible updates no longer flow as freely as they used to.

> Yes, having a Cargo warning for cases when there are dependency updates with incompatible MSRV would be a good feature.
> 
> As for security vulnerabilities, I think it should be solved by advisories like [RustSec](https://rustsec.org/) and by yanking affected versions. Also note that today crates can have minor (for pre-1.0, major otherwise) version updates with security fixes which do not get ported to older releases. So I don't think that the proposed MSRV-dependent version resolution is fundamentally different from the existing status quo in this regard.

---

Error messages

> IMO one of the big problems blocking this is the error messages. The current Resolver has pretty bad messages, it just reports on the last thing that did not work. So In effect the Resolver will just ignore all versions with a different MSRV leaving no record of why newer versions were not considered. This may be on the list of things to consider when/if we use [PubGrub](https://github.com/pubgrub-rs/pubgrub). But that is just my opinion.

> Wouldn't it be viable to implement an MVP with bad error messages first and work on improving them later?

---

rust-version as advisory only

> Is this actually a good idea? I feel like rust-version should solely be a safeguard, and not actually affect behavior. IMO pinning a Rust version without pinning dependencies is user error, and it's better for this to be a hard error instead of doing what might be the wrong thing.

> broken code doesn't help anyone. If cargo update / upgrade gives warnings that your rust version is too old to bring in all the latest updates then people will be more likely to run cargo upgrade as they are more likely to get a result that works. Pragmatically some stuff upgraded is better than nothing upgraded.

> But keeping "a conservative enough MSRV for your crate's audience" is not_possible if your upstream dependencies don't do the same. If upstream crates raise their MSRV without a major version bump, then you have little choice but to follow suit. As @leo60228 suggested you could pin those crates versions to before their MSRV bump, but that causes a different problem: Cargo refuses to build two semver-compatible versions of a single crate in a single build. That can result in downstream users' builds breaking, even if they don't care about MSRV. So Rust really does need an MSRV-sensitive resolver.

> In theory, this is true. In practice, I've found that upstream maintainers have usually been happy to accomodate my requests for changes to allow a more conservative MSRV -- just a matter of (a) being aware when changes happen and (b) engaging with upstream to voice your concerns/discuss trade-offs.

> One way to resolve this concern is to make the behavior opt-in. By default, behavior is at is today (modulo a better error message), but you can pass --rust-version flag to commands like update or generate-lockfile to request resolution at a specific version.

> Perhaps there could be an unstable option similar to cargo update -Z minimal-versions to perform a resolution that considers MSRV?

---

rustup interaction

> How should this interact with rust-toolchain.toml? At the moment it is only selecting a toolchain from rustup. Should it become the default version for update and generate-lockfile as well?

--- 


Other relevant discussions
- https://github.com/rust-lang/libs-team/issues/72
- https://github.com/time-rs/time/discussions/535

**Comment by epage** [comment]
## Blockers for MSRV in the resolver

The `rust-version` field existing in the Index so the resolver can reference it
- The more packages that get `rust-version` populated (which will happen over time as new crates are published), the more useful any features using it will be

## Problem

As an end-user, I might have rust install but haven't been keeping up with the latest and just want to `cargo install` without extra steps or might not be familiar with how to upgrade.  Even worse if I was using the distribution version.

As a maintainer, it is cumbersome to have an MSRV and for it to be something besides the latest.  When upgrading or pulling in a new dependency, a maintainer has to choose
- To forget it and update MSRV
- Try to get the depedency to hold back their MSRV which just shifts this burden to someone else
- Hand pick MSRV compatible versions, iterating with `cargo update --precise` until you are done
- Misuse version requirements without full understanding of the impact, e.g.
  - https://github.com/rust-lang/cargo/issues/6584#issuecomment-1308032517
  - https://www.reddit.com/r/rust/comments/p8clcx/how_to_fix_cargo_dependency_issue/

Similarly, as a corporate user, I might be aligned to specific rust versions that have been integrated into our build process or certified for our use.

As a maintainer, it can be tiring to deal with people making special requests for MSRV and I want things to "just work" so there is less reason for these requests to be made
- Sometimes these requests don't come with any requirement more than "someone else asked me"
- The conversations in [libc](https://github.com/rust-lang/libs-team/issues/72) and [time](https://github.com/time-rs/time/discussions/535) MSRV policy discussion don't feel all that unique

## Proposed behavior

By default, we resolve with MSRV

Implementation: the resolver would treat `rust` as just another dependency, one that exists in a different namespace as packages. 

For local development tasks
- We effectively treat it as all rust versions exist.
- workspace members with a `package.rust-version` effectively have `rustc.version = "<=package.rust-version"`
- non-workspace members with a `package.rust-version` effectively have `rustc.version = "^package.rust-version"`

For `cargo install` (#10903)
- For unconstrained versions, select a version of the package compatible with `rustc -V`
- If `--locked` or `--ignore-rust-version` are not set:
  - Either
    - We only act as if `rustc -V` exists
    - The current toolchain is another node in the dependency graph with the dependency `rustc.version = "=package.rust-version"`
  - all packages with a `package.rust-version` effectively have `rustc.version = "^package.rust-version"`

For `cargo add`, versions newer than the selected package's `package.rust-version` will be ignored (#10653)

`--ignore-rust-version` would be update to opt-out of the MSRV resolver

## Challenges with doing MSRV-aware resolver with possible solutions

Error reporting
- The proposed implementation should have errors that are no worse than existing resolver errors
- Heuristics could possibly be added that don't say what caused the failure but provide context for the user to decide what caused the failure (e.g. reporting versions of a dependency that are yanked or have too high of an MSRV)
  - This is likely over simplifying and making it sound easier than it is

Users running stale, possibly insecure versions
- Note: this is already a problem with semver incompatible versions and renames
- `cargo audit`, `cargo deny` exist but require users to opt-in
- Github will notify of vulnerabilities through Dependabot, even using the push-message to tell users on the command-line but this is limited to Github
- `cargo update` report upgrades/downgrades in a table with old, new, latest compatible, latest incompatible versions
  - Make `--verbose` report the status for all crates not on latest version
  - A `--dry-run` flag is already available to see the status without upgrading
- `cargo add` and `cargo install` could warn when not selecting the latest version of the specified package, informing the user what the latest compatible and/or incompatible version is
  - This does leave off any notifying the user for any indirect dependencies
- Longer term, we can explore integrating `cargo audit` into cargo (e.g. #10016) which will help with both MSRV incompatible and semver incompatible versions

Upgrading to a new rust version causes my lockfile to change
- This is particularly bad for CI jobs that ensure the `Cargo.lock` is aligned with `Cargo.toml`
- This is only a problem if...
  - The version of a dependency already has a `rust-version` entry in the index
  - The dependency's `rust-version` is newer than your `rust-version`
  - You do not have a CI job to verify MSRV
- A user can resolve this on one-off basis by updating the lockfile
- So far, this does not seem like a big enough concern.  If people find a reason to raise the priority, we could make this opt-in until the next edition after which it would be opt-out

Maintainers who want to different rust versions between MSRV verification and regular development
- Use case: As a binary author, I want my pre-built binaries to have all the latest features/fixes but I want an older MSRV for compatibility with Debian users
- Maintainers can use the `--ignore-rust-version` flag to opt-out of the MSRV-aware resolver and maintain multiple lockfiles
  - We might need to add `--ignore-rust-version to more commands
  - While this is more of a burden, I suspect this is narrow enough of a use case to be acceptable
  - To make this more ergonomic, we could add a `rust-version` config field that overrides the effective MSRV with either a "<x>.<y>.<z>` or "package" (take from Cargo.toml) or `current` (`rustc -V`) so users can set an env variable or drop a `.cargo/config.toml` file

## Alternatives

Make MSRV-aware resolver opt-in
- MSRV jobs could use the "`--resolve-with-msrv`" flag
- MSRV build errors could suggest passing the "`--resolve-with-msrv`" flag
- 🔴 This steers the community away from having a cohesive, dependable MSRV experience like we do for semver.
  - I attribute the dependability of semver within the rust ecosystem to (1) it was there from the beginning, (2) its encoded as the default in our tools with version constraints, and (3) the lack of `Cargo.lock` for libs has us self-policing each other for semver violations
  - By not making MSRV the default resolver policy, its putting in extra roadhumps to maintainers doing the right thing, making it less likely the ecosystem will have a good MSRV story (`package.rust-version` actually set and is accurate)
- 🔴 For new or irregular users, this creates hurdles when starting a task (e.g. `cargo install`).  If enough of these hurdles are present, it could sour the person on their experience with Rust.

**Comment by jonhoo** [comment]
I'd like to propose that `cargo update` should also always warn if an update is withheld due to MSRV. I think there's a decent argument that it should even hard-error unless a flag is passed, but that _may_ be overkill.

**Comment by epage** [comment]
> I'd like to propose that cargo update should also always warn if an update is withheld due to MSRV

What if the table mentioned this?  While I didn't specify it, that was one of my expectations with the table to report why a newer version may not be available

>  I think there's a decent argument that it should even hard-error unless a flag is passed, but that may be overkill.

If the flag always has to be passed in, it loses its meaning and instead only serves as a barrier which just instills a feeling of frustation.  For any project lagging behind its dependencies, this will always be needed.

**Comment by jonhoo** [comment]
> > I'd like to propose that cargo update should also always warn if an update is withheld due to MSRV
> 
> What if the table mentioned this? While I didn't specify it, that was one of my expectations with the table to report why a newer version may not be available

Maybe this is a "me" thing, but I generally do not want verbose output from commands. A table would be far too noisy for me, and I would only want one if I explicitly passed in `--verbose`. Ideally, in my mind, `cargo update` would print nothing, with an option to print a list, and another option to print a full table. But this warning should be surfaced no matter which output mode is used.

> > I think there's a decent argument that it should even hard-error unless a flag is passed, but that may be overkill.
> 
> If the flag always has to be passed in, it loses its meaning and instead only serves as a barrier which just instills a feeling of frustation. For any project lagging behind its dependencies, this will always be needed.

I think my proposal here stems from the fact that I worry a _lot_ about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result. To the point that I think having MSRV-aware resolving by default for `cargo update` is a mistake. I fully acknowledge there's a trade-off here, but I think valuing "Cargo.lock represents `rust-version`" higher than "you're using the latest possible version of your dependencies" is going to come back to bite us over and over again. *If* we go that path though, I think at the very least the decision to willingly take older versions should be made very explicit.

**Comment by newpavlov** [comment]
>I think my proposal here stems from the fact that I worry a lot about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result.

Following this logic Cargo should not use Cargo.lock at all and instead always fetch dependency updates on each build job. It's obviously not how we work with dependencies, because updating them has its own set of risks.

**Comment by rpjohnst** [comment]
The difference is that you can currently expect to get the latest versions compatible with what *you* wrote in `Cargo.toml` when running `cargo update`. With MSRV-dependent resolution, you might start silently getting older versions based solely on the *libraries'* `Cargo.toml`s and your current version of Rust.

And "your current version of Rust" is not always a real constraint- sometimes you just haven't run `rustup update` yet, and would be perfectly willing to do so if you knew it would help.

**Comment by jonhoo** [comment]
> >I think my proposal here stems from the fact that I worry a lot about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result.
> 
> Following this logic Cargo should not use Cargo.lock at all and instead always fetch dependency updates on each build job. It's obviously not how we work with dependencies, because updating them has its own set of risks.

I don't think that follows. `cargo update` is an explicit operation I run to update. Cargo does not run it on my behalf.

That said, this is related to one of the concerns I know have been raised in response to the suggestion that folks should start checking in Cargo.lock for libraries — that doing so would mean less automatic testing of new versions of dependencies, and that such an outcome would be detrimental to the health of the ecosystem.

**Comment by newpavlov** [comment]
@rpjohnst 
>With MSRV-dependent resolution, you might start silently getting older versions based solely on the libraries' Cargo.tomls and your current version of Rust.

Not silently. With an appropriate warning message on `cargo update` or on first generation of Cargo.lock.

@jonhoo 
>I don't think that follows. cargo update is an explicit operation I run to update. Cargo does not run it on my behalf.

By having Cargo.lock in your workspace (regardless whether it committed to VCS or not) you are using "old versions of their dependencies", unless you explicitly run `cargo update`. If we are consider it's fine today, what exactly MSRV-dependent resolution changes? If anything, the current status quo is even worse than the proposed warning system.  Today you do not get any built-in warnings if there are outdated dependencies in your tree, you have to use separate tools like `cargo outdated`.

**Comment by tarcieri** [comment]
Per the [summary above](https://github.com/rust-lang/cargo/issues/9930#issuecomment-1489089277) it seems like there's fairly widespread agreement that using an older version to be MSRV compatible should print a warning

**Comment by illicitonion** [comment]
> I think my proposal here stems from the fact that I worry a _lot_ about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result.

In general I agree with this.

> To the point that I think having MSRV-aware resolving by default for `cargo update` is a mistake. I fully acknowledge there's a trade-off here, but I think valuing "Cargo.lock represents `rust-version`" higher than "you're using the latest possible version of your dependencies" is going to come back to bite us over and over again.

It feels like there's a missing piece here around intent.

If we consider a "how-much-I-care-about-msrv" equivalent to Cargo.toml's existing "maintenance" field, I can imagine the following values:
* I really strongly care about MSRV
* My real aim is to target e.g. current minus 6 months, so I'd be happy to bump to whatever that is
* I would be happy to bump this if there's any good reason to
* I don't actually care about this at all, but cargo started setting a default value for me so here it is

And if we consider the operations a maintainer may be doing on a crate, I can imagine being in any of the following modes:
* I'm just trying to work keeping to my original MSRV
* I'm trying to work, and am happy to explore changing my MSRV if it's convenient
* I'm explicitly looking to bump my MSRV

When performing operations, particularly `cargo update`, these two factors combine to give reasonable behaviours.

If you really strongly care about MSRV, you probably want an explicit flag to ever violate MSRV. If you actually have a more flexible policy, maybe a different default makes sense. But it feels like being able to express the intent of _why_ this is your MSRV may affect how cargo should act.

> you might start silently getting older versions based solely on the _libraries'_ `Cargo.toml`s

It feels like _transitive_ dependencies' MSRVs should be capped by the actual package being operated on from the command line is. 

If I have a package `foo` which specifies MSRV 1.56, and I have a dependency which specifies MSRV 1.54, any new package I consider bringing in should be capped by 1.56 not 1.54. I think in the latest proposal, this is actually slightly ambiguous as to which crates in the dep graph factor into MSRV calculation.

**Comment by jonhoo** [comment]
> By having Cargo.lock in your workspace (regardless whether it's committed to VCS or not) you are likely using "old versions of dependencies", unless you explicitly run `cargo update`. If we consider it's fine, then what exactly MSRV-dependent resolution changes? If anything, from your point of view, the current status quo should be even worse than the proposed warning system. Today you do not get any built-in warnings if there are outdated dependencies in your tree, you have to use separate tools like `cargo outdated`.

The big difference is that running cargo update is an indication of intent. When I run that command I'm saying "I want to update now", and it's unfortunate that after running that command I'm not actually up to date.

**Comment by newpavlov** [comment]
>it's unfortunate that after running that command I'm not actually up to date.

As mentioned in my previous comment, it's quite likely you are not "up to date" even today. One of you dependencies could have a breaking release and you will have absolutely no idea about it without running a tool like `cargo outdated`. And even before that, the fact that you are not on the latest toolchain already makes the talk about being "up to date" moot.

**Comment by jonhoo** [comment]
It's true that major version updates require more work, but that's not an argument that we should make the situation (significantly) worse by extending the same problem to minor version updates as well.

I agree it's a problem when someone's Rust version is behind, but a) it's a smaller problem, and b) the proposal here won't surface that any more than the current state of affairs — quite to the contrary.

**Comment by 8573** [comment]
> With MSRV-dependent resolution, you might start silently getting older versions based solely on the _libraries'_ `Cargo.toml`s and your current version of Rust.

Is this true?  I had thought that "MSRV-dependent resolution" would depend on the `rust-version` declared in one's `Cargo.toml`, not on the Rust version one happens to be using to build.

**Comment by djc** [comment]
> > > I'd like to propose that cargo update should also always warn if an update is withheld due to MSRV
> > 
> > What if the table mentioned this? While I didn't specify it, that was one of my expectations with the table to report why a newer version may not be available
> 
> Maybe this is a "me" thing, but I generally do not want verbose output from commands. A table would be far too noisy for me, and I would only want one if I explicitly passed in `--verbose`. Ideally, in my mind, `cargo update` would print nothing, with an option to print a list, and another option to print a full table. But this warning should be surfaced no matter which output mode is used.

+1. A table (assuming your proposal would be similar to the table which appears for `cargo upgrade` today) scales really badly to larger workspaces (as I've explained before for `cargo upgrade`). Instead, output should focus *only* on deviant/attention-worthy cases.

> > > I think there's a decent argument that it should even hard-error unless a flag is passed, but that may be overkill.
> > 
> > If the flag always has to be passed in, it loses its meaning and instead only serves as a barrier which just instills a feeling of frustation. For any project lagging behind its dependencies, this will always be needed.
> 
> I think my proposal here stems from the fact that I worry a _lot_ about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result. To the point that I think having MSRV-aware resolving by default for `cargo update` is a mistake. I fully acknowledge there's a trade-off here, but I think valuing "Cargo.lock represents `rust-version`" higher than "you're using the latest possible version of your dependencies" is going to come back to bite us over and over again. _If_ we go that path though, I think at the very least the decision to willingly take older versions should be made very explicit.

I wonder if there's room for nuance based on `update` (semver-compatible only) and `upgrade` (potentially including semver-incompatible) semantics. Arguably `update` means "give me the latest version of everything as long as it is compatible with my current dependency specifications", and `update` today doesn't warn about the availability of newer, semver-incompatible versions. In this sense, `cargo update` already doesn't mean "you're using the latest possible version of your dependencies".

**Comment by epage** [comment]
> I think my proposal here stems from the fact that I worry a _lot_ about folks using old versions of their dependencies, and ending up with insecure, buggy, or under-performing software as a result. To the point that I think having MSRV-aware resolving by default for `cargo update` is a mistake. I fully acknowledge there's a trade-off here, but I think valuing "Cargo.lock represents `rust-version`" higher than "you're using the latest possible version of your dependencies" is going to come back to bite us over and over again. _If_ we go that path though, I think at the very least the decision to willingly take older versions should be made very explicit.

I can understand the concerns over stagnation, security, and "do what I told you" behavior but I also feel we need to meet people where they are at.  A workflow that is doomed to fail on first try and requires a special flag to get working, even if you are told about that flag, is a frustrating experience.  Then this compounds further when you are maintaining an MSRV so everything you do, you need to make sure you don't accidentally get too new versions without passing in a special flag.  Within that workflow, it is reasonable to run `cargo update` and only get what will work for my project.  It doesn't update dependencies that break semver.  It also doesn't upgrade dependencies that something has `=` on.

**Comment by jonhoo** [comment]
Let me try to be a bit more explicit here about my beliefs, assumptions, and wants.

I think an outcome where the Rust ecosystem is generally less up-to-date over time is significantly worse than it being painful to maintain MSRV. I also think it affects significantly more people.

I think making `cargo update` respect `rust-version` by default is going to lead to that outcome.

I think there are (many) more application authors than library authors, and thus that most users do not need to care about MSRV or `rust-version` as far as their own crate goes. And thus it's fine to impose some cost for MSRV checking since it affects fewer users (and per the previous point I think the downsides to the majority of users is significant).

I think `cargo update` should warn if it _ever_ doesn't update to the latest version to highlight that the caller may want to take another action to unblock the update.

Consequently, I think it's a mistake that it does not warn today in the presence of `=` or `<` specifiers if they prevent an update.

I also think it's a mistake that it does not warn today in the presence of newer major versions. I think it's something that's going to come back to bite the entire Rust ecosystem, as it leads to stagnation without visibility, which in turn leads to everything slipping further and further behind (dependabot is not sufficient to solve this problem).

I think it _may_ even be reasonable to fail if a `cargo update` cannot update to the newest available version of  something (but still update `Cargo.lock` as best it can) to clearly indicate to users that they are _not_ up to date.

On the topic of verbosity, I think the standard output of a command should be concise, and only point out deviations. That means no tables and only "anything that didn't do the expected thing" such as if an update didn't happen. I think a table view is totally reasonable, but should be opt-in, not opt-out. I also think that the table view should only include "uncommon" rows (e.g., those were an update was partial) by default. I hold the first of these views weakly, and the second strongly.

**Comment by remram44** [comment]
I don't understand, wouldn't it be easy for application author to not specify an MSRV at all? In this situation, do they run into any problems?

**Comment by Nemo157** [comment]
> Hence, I think the presence or absence of the `rust-version` key in the current crate's manifest is a pretty good indicator of intent in both situations.

> * The more packages that get `rust-version` populated (which will happen over time as new crates are published), the more useful any features using it will be
> &mdash; <https://github.com/rust-lang/cargo/issues/9930#issuecomment-1489269471>

These two comments seem to be at odds, my understanding of @epage's implication is that eventually _every_ crate should have a `rust-version` set so that downstream MSRV-following users can use that when picking their dependencies, even if those upstream crates aren't themselves following an MSRV and just increasing `rust-version` on every release. (Unless the intent is that you create a new branch for every release and set `rust-version` just for that published commit, or that eventually `cargo package` automatically sets it; but IMO the first is too much effort and the latter hasn't been proposed as an actual change yet).

Really an MSRV-following application author is the one that _doesn't_ need to publish their `rust-version`, they only need to know it transiently during `cargo update`. It's all the libraries they use that need to have that metadata published for them to then consume during version resolution.

**Comment by adamreichold** [comment]
> even if those upstream crates aren't themselves following an MSRV policy and just increasing rust-version on every release. 

While it is not a very conservative one, it still is an MSRV policy that the solver could use as an additional constraint. That said, personally, I would not add a `rust-version` property to a library that does not really have an MSRV policy, because that property needs to tested for and considered when making code changes.

> Really an MSRV-following application author is the one that doesn't need to publish their rust-version, they only need to know it transiently during cargo update. It's all the libraries they use that need to have that metadata published for them to then consume during version resolution.

They don't need to publish it. But they don't need to publish their Cargo manifest at all. (I don't think `cargo install` is a good long-term solution for distributing applications.) Still, the presence of the key in their Cargo manifest can signal the intent of following that MSRV to the dependency solver which should then consider it.

> I as a conscientious maintainer that wants to support MSRV restricted downstreams have rust-version = "1.57" set in my Cargo.toml because that was the latest version at the time of my last release. I now am developing a new feature that needs a new dependency so I cargo add foobar. Despite there being a foobar = "2.0.0" available to add, that has rust-version = "1.59", so instead cargo decides to add foobar = "1.10.2". Unless I then run cargo outdated or something I won't realise that there's a newer version available and that I need to increase my rust-version to allow access to it.

My first guess would be that the manual step of running `cargo add foobar` should clearly spell out the reasoning behind its choice of version `1.10.2`, so that I am aware that my MSRV prevents me from using version `2.0.0`. I can then make a conscious decision to bump my MSRV if I deem the difference between these version a sufficient reason for doing so.

As a general remark, I suspect that it will not be possible to resolve the difference of approach between following the latest versions available and constraining one's dependency to give "MSRV closure". I do think there is room for both approaches though and hence I think the main usability issue to decide is how I as a maintainer can opt into the "slow world". I think the initial status quo of not having `rust-version` set, could default to the "fast world", because as written above keeping an MSRV entails more than just handling the dependency closure even though that is the most chore-like part of it.

**Comment by djc** [comment]
Note that an `edition` imposes a `rust-version` constraint as well: for a crate that has `edition = "2021"`, `rust-version` is set to 1.56. So from that perspective, if you use `cargo init` today your crate will have a `rust-version` even though it is not explicitly mentioned in the manifest. From this perspective, it seems likely that we will trend towards all crates having a `rust-version` set, but it does make it harder to derive the author's intent about whether they want their crate's dependencies to match a particular MSRV. RFC 2495 being a minimal implementation, the **only** goal its implementation serves is to help consumers decide whether the crate is fit for compilation using a particular Rust version. If you want to communicate intent about which MSRV a transitive dependency graph *should* support, the implementation has to change (or we should additional metadata to the manifest).

> I think there are (many) more application authors than library authors, and thus that most users do not need to care about MSRV or `rust-version` as far as their own crate goes. And thus it's fine to impose some cost for MSRV checking since it affects fewer users (and per the previous point I think the downsides to the majority of users is significant).

This makes a lot of sense to me.

> I think `cargo update` should warn if it _ever_ doesn't update to the latest version to highlight that the caller may want to take another action to unblock the update.
> 
> Consequently, I think it's a mistake that it does not warn today in the presence of `=` or `<` specifiers if they prevent an update.
> 
> I also think it's a mistake that it does not warn today in the presence of newer major versions. I think it's something that's going to come back to bite the entire Rust ecosystem, as it leads to stagnation without visibility, which in turn 
leads to everything slipping further and further behind (dependabot is not sufficient to solve this problem).

Today, `cargo update` already prints a list of actions it took (added dependencies, upgrades and downgrades -- though downgrades are shown as upgrades in today's stable Rust). Conceptually it would make sense to me to add lines of output to it that mention if a newer version is available but unable to be used because of version constraints in the compilation target.

> I think it _may_ even be reasonable to fail if a `cargo update` cannot update to the newest available version of something (but still update `Cargo.lock` as best it can) to clearly indicate to users that they are _not_ up to date.

I'm not sure I consider this to be reasonable. So here you are referring to changing the process' return value in that scenario (and it's output)? I'm not sure I agree that this would be reasonable.

> On the topic of verbosity, I think the standard output of a command should be concise, and only point out deviations. That means no tables and only "anything that didn't do the expected thing" such as if an update didn't happen. I think a table view is totally reasonable, but should be opt-in, not opt-out. I also think that the table view should only include "uncommon" rows (e.g., those were an update was partial) by default. I hold the first of these views weakly, and the second strongly.

Do you consider `cargo update`'s current output a "table view"? IMO its current output is useful, and I don't think it should be more concise by default.

<img width="575" alt="Screenshot 2023-04-12 at 13 40 37" src="https://user-images.githubusercontent.com/158471/231446370-5d902c06-c09d-48c4-8cc7-54942e6738c6.png">

**Comment by adamreichold** [comment]
> Note that an edition imposes a rust-version constraint as well: for a crate that has edition = "2021", rust-version is set to 1.56. So from that perspective, if you use cargo init today your crate will have a rust-version even though it is not explicitly mentioned in the manifest. From this perspective, it seems likely that we will trend towards all crates having a rust-version set, but it does make it harder to derive the author's intent about whether they want their crate's dependencies to match a particular MSRV.

Isn't it purely an implementation issue to make Cargo's internal data model differentiate between an implied MSRV due to an edition being set and a `rust-version` actually specified in the manifest file thereby modelling the intent represented by the crate author explicit setting a `rust-version`?

**Comment by tarcieri** [comment]
Alternatively to inferring it from `rust-version`, perhaps there could be an explicit setting in `Cargo.toml` or `.cargo/config.toml`?

Something like `resolver-msrv = "1.56"` perhaps?

I'm also curious whether it would make more sense for it to be something that happens at the granularity of a workspace, rather than considering the individual packages within the workspace.

**Comment by epage** [comment]
@jonhoo and I got to talk synchronously about this

Some quick things we can get out of the way before the details
- @jonhoo fully agrees with MSRV resolution for `cargo install` using the current Rust version
- @jonhoo cares less about how `cargo update` is rendered and more that the output shouldn't be overwhelming (ie he'd rather have no output)
  - He'd prefer any MSRV or version incompat warnings to be coalesced to "NUM packages are behind because of X" than enumerating each.  We could recommend `cargo update -p <package>` or `cargo update --verbose` to see a non-coalesced message

So moving on to non-cargo-install cases...

Some related threads
- The cargo team has been talking internally about whether to change the default policy for committing `Cargo.lock` for `[lib]` crates.
  - Its nearly impossible to test MSRV without doing this today and the hoops of maintaining a `Cargo.lock.msrv` just don't feel worth it
  - This makes it hard to do things like `git bisect` to identify the source of bugs
  - Not having a `Cargo.lock` file can be a source of red-master which can be a frustrating experience for a contributor to a project (as they assume the problem is their own and waste time digging into it, it then blocks them from progress)
  - But by not having a `Cargo.lock`, we effectively get a distributed crater run on the whole ecosystem, helping us more quickly identify bugs / breaking changes and getting them fixed by exerting the pressure of the ecosystem-wide pain
    - But this is limited to `[lib]`s and its more likely the `[[bin]]`s that we care about
- #5657
  - **See [my comment](https://github.com/rust-lang/cargo/issues/5657#issuecomment-1515275965) (seriously, read it; most of it applies here and I'm not going to re-tread it)**

People's Rust version might be behind because
- A company in a safety critical domain might standardize on specific compiler versions and upgrading takes extensive testing to protect against cases like miscompilation or subtle changes in how floating point numbers are moved to and from the FPU (64-bit <-> 80-bit conversions losing precision)
  - This matters most in the safety critical parts o the code base but not everything is safety critical, so dependencies might be a bit more free
- A Debian user trying out Rust, installing via `cargo install`, or getting Rust code indirectly like with Python packages
- A developer who casually uses Rust every once in a while and might fall behind from neglect but doesn't want to be forced to update every time they do something
  - We've seen a smaller version of this with `cargo` pulling down the index and these casual users being hit with "Resolving deltas" every time they touch Rust.
  - **Note:** We have a selection bias in our conversations for those who are passionate / involved / have enough time to participate, making us more likely to also be aware of and updating to new releases.  This is the opposite of this type of user

We also want a culture that moves forward rather than a culture of stagnation, which is affected by how tools affect our behavior

Problems with stagnation include:
- People running older software with known security vulnerabilities because they are held back by MSRV, breaking changes, etc
- Big, flag-day upgrades are expensive.  Its better to pay the compiler and dependency upgrade cost incrementally.
  - The more the gap grows, the more likely it is to grow because it takes more justification to pay down this technical debt.  
  - Most migration guides are written assuming incremental upgrades because there are too many variations of large upgrades to account for
- The community being in the C89 / C99 / C++98 trap of never getting to use new compiler features which we are most likely to evolve to as adoption grows if we do not provide a smooth workflow for people who have a reason to be on older compilers (the future community's pressure to conform will push us back to older versions, just look at rust-lang/libs-team#72)

Tied into the [two main workflows](https://github.com/rust-lang/cargo/issues/5657#issuecomment-1515275965) is the idea of what your target rust-version is  (whats used for development, official binary builds) vs your verified minimal rust-version

Possible solutions we discussed (which are similar to [my other comment](https://github.com/rust-lang/cargo/issues/5657#issuecomment-1515275965), see it for some more trade offs)
- ~~For local development, resolve to `rustc -V`~~
  - This is out because `Cargo.lock` should be resolved in a stable fashion to the point that being smarter about what dependencies are included requires a version bump (see #11938)
  - Without a rust-toolchain file, users might be on different rust versions.  Even a single user might be on multiple versions if they are testing out issues
- Resolve to [`rust-toolchain.toml`](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file)
  - This expresses the intent of your target rust-version
  - But this couples cargo to rustup which might not even be used though you could say it could be a general configuration file 
  - Can't always extract a version from it, depending on how its being used (e.g. with a `path`), so we'd need to detect the file and run `rustc -V` if it exists
    - What is the ceiling dir for what we should check for affecting the resolver?  If its out of the workspace?  out of the repo?
  - Does this actually work in all of those corporate environments?
  - For those using it with `package.rust-version`, this becomes redundant information that needs to be kept in-sync
- Resolve to `package.rust-version`
  - This **is** the user signalling to use they have an MSRV and want to follow it.  Any developer coming to that project with a version at or newer than MSRV should be able to contribute
- Opt-in CLI flag
- Sticky opt-in CLI flag (setting stored in `Cargo.lock`)

Regarding security: being on an old version doesn't inherently make you less secure.  For unreported vulnerabilities, the main risk is there are known unreported vulnerabilities.  Otherwise, we could also improve the security angle by integration cargo-audit behavior into cargo (#7678, #10016)

Regarding "ensuring my dependencies do not break my dependents"
- If a `Cargo.lock` file is not committed on either end, that only means you are using the versions as of the time that CI ran
- If `Cargo.lock` is committed, it only tells you that the version in that `Cargo.lock` worked, not the versions used by your dependents

I know there is more but I've forgotten it being 24-hours later...

**Comment by jonhoo** [comment]
I mostly agree with the summary above. I want to clarify my thoughts on one point though, which is where I think the disagreement about where we should land in the trade-off space of this decision comes to:

> Regarding security: being on an old version doesn't inherently make you less secure. For unreported vulnerabilities, the main risk is there are known unreported vulnerabilities. Otherwise, we could also improve the security angle by integration cargo-audit behavior into cargo

I think this is false. In general, newer versions have fewer bugs than older versions, whether they are known or not, and whether they are security-related or not. By running older versions of a dependency, you're more likely to have either active or latent security risks, performance issues, or even just outright incorrect behavior. Whether that _matters_ depends on where you're building. If you're building in a development context, I totally agree that most of these are fine, and using older versions is A-OK. But if you're building for any kind of production, in my opinion, you should strongly favor using the newest stable versions of all your dependencies that your environment allows (whether rustc or crates). It follows from this that if your current environment prevents you from using the newest stable versions, you should be alerted to this fact and given an opportunity to update the environment rather than have the build "just work" by falling back to older versions.

Unfortunately, there isn't a way in Cargo to cleanly distinguish between "during development" and "for production". `cargo install` is the closest we have, but it's both used for development and not always used for production. Which then leaves us in a position where we need to choose: does Cargo seamlessly use old versions, or push you towards updating. I think the argument is in favor of the latter, but I completely acknowledge that there's a trade-off involved where there isn't (and won't ever be) universal consensus.

Ultimately, I took from my chat with @epage that where we disagree is on where we should come down on the following trade-off:

1. A user that checks out, builds, modifies, and submits changes to a crate should encounter zero Rust version related friction in doing so, unless their version is older than `package.rust-version` or they explicitly try to write new code that requires a newer Rust version.
2. When I build a crate for production, assuming there isn't a `Cargo.lock` or I have run `cargo update`, I get the newest version of my dependencies that my `rustc` allows.

The logic for how we get to these being deeply at odds with one another is a little convoluted, but I believe they are. That said, if someone has a solution that doesn't force these two to be traded off against each other, please do speak up!

If that is indeed the trade-off we have to make (or the other options are to make _worse_ trade-offs), my instinct is to favor the second one. It makes me very sad to not be able to make the contributor experience for those with older `rustc` entirely friction-free, but I think the cost of 2 is too great to trade that one off. I know @epage comes down on the other side of that, and would prefer path 1.

Ultimately, I'm not a decision-making authority here. I think the Cargo team will have to collectively decide which of these use-cases is the one they want to favor over the other, and either way they'll be trading something away.

**Comment by epage** [comment]
> I think this is false. In general, newer versions have fewer bugs than older versions, whether they are known or not, and whether they are security-related or not.

I am unsure if we can make such a statement.  So long as new development is happening, new bugs are being introduced, including bug fixes (see #11999 for a recent example).  We are helped by tests but that mostly helps us avoid repeating the same mistakes over again.  We're human; we are creating in finding new mistakes to make.

> If that is indeed the trade-off we have to make (or the other options are to make worse trade-offs), my instinct is to favor the second one. It makes me very sad to not be able to make the contributor experience for those with older rustc entirely friction-free, but I think the cost of 2 is too great to trade that one off. I know @epage comes down on the other side of that, and would prefer path 1.

One of the reasons I lean the other way is my assumption that
- For most projects, (1) is likely sufficient in all cases and are less likely to put in time and resources to a lockfile dance
- (2) is most likely needed for Big Projects which have more people and processing resources to put towards doing a lockfile dance

For context, how I'm managing my personal projects
- All projects have a lockfile checked in, regardless of whether its a `[[bin]]` or a `[lib]`
- All dependencies are kept within MSRV
- A best-effort is made to imitate `-Zdirect-minimal-versions`
- `[lib]`'s automated upgrade runs monthly and only upgrades across breaking releases
  - If I need a new feature, I upgrade manually
- `[[bin]]`s automated upgrade runs monthly and updates all dependencies to latest
- *(in-development)* all projects are getting an automated upgrade of `package.rust-version`.

**Comment by adamreichold** [comment]
> Opt-in CLI flag

Coming from a "maintaining a library targeting Debian's Rust toolchain" background, I think an opt-in flag to produce an MSRV-compatible `Cargo.lock` file (based on `package.rust-version`) would be a minimum viable solution that could significantly simplify our work by replacing kludges like  

https://github.com/PyO3/pyo3/blob/2f8bf513b506e8b093794ad695b2d6628e6ff8d2/noxfile.py#L408-L508

without significantly impacting the status quo of version resolution.

(I think this reasoning applies to `-Zdirect-minimal-versions` as well, i.e. producing a compatible lock file upon request is enough to simply and automatically make CI jobs work would be a big improvement without impacting any defaults.
Usually, our contributors have more issues with our CI than just dependency resolution, e.g. we test additional Python versions and variants as well as additional architectures. We do try to mitigate this by making it feasible to run the same build automation locally and often maintainers push build fixes onto contributor PR to avoid putting this extra work onto the contributor's shoulders.
Manually juggling MSRV-compatible versions is currently a relatively common occurrence of such fix-ups.)

**Comment by jonhoo** [comment]
> So long as new development is happening, new bugs are being introduced, including bug fixes

While it's true that new bugs arise over time as well, the rate of bugs being fixed should be higher than the rate at which bugs are introduced for any given project, especially when it comes to already-existing features (as opposed to new feature development).

**Comment by remram44** [comment]
> the rate of bugs being fixed should be higher than the rate at which bugs are introduced for any given project

If that were true there wouldn't be any bugs. Projects are born with no code and no bugs.

**Comment by jonhoo** [comment]
They're also born with no features. Hence the carve-out in my statement that this doesn't apply to new feature development. Over time though, the rate of new features in a project tends to go down (or when they are introduced, most users mainly rely on the old features).

Anyway, this is starting to veer off-topic. My point was simply that I think newer-versions-are-better is a fairly true statement. There are _instances_ where it's not true, but on balance it's more true than not. And I would contend by a solid margin, especially if you're comparing with a quite-old version (say, one that builds with a 2 year old Rust compiler).

**Comment by epage** [comment]
rust-lang/crates.io#6267 is now in [production](https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/rust-version.20field)!

**Comment by epage** [comment]
[My comment](https://github.com/rust-lang/cargo/issues/9930#issuecomment-1489269471) is now hidden by default.

**Comment by epage** [comment]
killercup/cargo-edit#851 applies the coalescing idea to cargo-upgrade

CC @djc

**Comment by YellowOnion** [comment]
Does this fix cargo not making a backwards compatible lock file? I've been banging my head against the wall trying to figure out how to generate a lock file that works with older versions.

**Comment by epage** [comment]
@YellowOnion could you clarify what you mean by a backwards compatible lockfile?

For myself in reading that, I would consider that to apply to the format and not the dependencies picked for that format (which is what this issue is about).  #12861 is exploring adding more checks on the format we use.

**Comment by YellowOnion** [comment]
@epage I have rust-version = 1.65 in the toml of bcachefs-tools, but cargo 1.69, will generate a lock file that won't even compile without 1.70, I spent like 6 hours trying to patch latest bcachefs-tools to compile on my OS's default rust, cargo and rust-analyzer just hates anything older than 1.7.

**Comment by epage** [comment]
Sounds like the file format is backwards compatible but the dependencies selected are not compatible which is what this issue addresses. If you install a nightly, you can run `cargo generate-lockfile -Zmsrv-policy`.

Is there a reason you are developing on 1.69 rather than the latest rust or your MSRV?

**Comment by YellowOnion** [comment]
@epage I'm not developing, I'm packaging it, and the OS only has 1.69 in the package manager / CI system, next stable release is this month sometime, surprised I can't use a ~5 month old cargo without things going wrong.

**Comment by epage** [comment]
Thanks for the explanation! Its useful to know what cases people are running into problems as it helps us better understand solutions, like the work I'm doing on the pre-rfc for this: https://internals.rust-lang.org/t/pre-rfc-msrv-aware-resolver/19871

**Comment by zh-jq-b** [comment]
Is it possible to only respect the `rust-version` of bin crates in a workspace?
We have several lib crates in our workspace that doesn't need to use new features so their rust-version field may be 1.66, 1.70, 1.74 or 1.80, and the bin crates will have a much higher MSRV requirement such as 1.80, is it possible to generate the Cargo.lock file matches MSRV 1.80 automatically (by default)?

**Comment by zh-jq-b** [comment]
> At this time, no. However, the version that is being stabilized tomorrow takes into account all MSRVs when selecting a version, picking the most compatible package within a version requirement. If the dependencies for a bin have a high enough version requirement that it excludes versions for 1.74, it will then try to be compatible with 1.80.

This will work for me I think. I will try it after available in nightly. Thanks

> One of the future possibilities is to resolve with a specific version in mind. Your use case though is a bit different than the ones we explored previously. Usually its people wanting MSRVs to be satisfiable during development or wanting all latest dependencies. How do you validate the libraries? Why go for the middle ground rather than making all MSRVs work or developing with latest?

We have a workspace level MSRV, which is also the bin crates' MSRV, we will keep them unchanged in LTS branch, so a Cargo.lock with respect to this MSRV is useful to us.
The lib crates are different as they may be published someday to be used by others, and I don't want to bump the MSRV as long as they didn't use any new features, so each of them has it's own MSRV policy.

**Comment by epage** [comment]
Are you verifying the lib crate MSRVs?

**Comment by zh-jq-b** [comment]
> Are you verifying the lib crate MSRVs?

No

**Comment by Kixunil** [comment]
I have a similar use case. A library crate with conservative MSRV and a binary crate with a higher one. The binary is basically a fancy example of the library. I don't remember the exact reason why I opted for higher MSRV in case of example but it might have been that I just didn't want to manually resolve the additional dependencies (complex tree involving HTTP, tokio...) and the example is not really that important. :)

**Comment by Pr0methean** [comment]
This still requires that every dependency, even if it's not enabled, have a version that matches the whole project's MSRV. So it won't fix all the problems that would be fixed by also letting features have MSRVs of their own.

**Comment by ijackson** [comment]
I have a similar problem to https://github.com/rust-lang/cargo/issues/9930#issuecomment-2579248152 https://github.com/rust-lang/cargo/issues/9930#issuecomment-2605528591.

For now I have changed our project process documentation to suggest using `--ignore-rust-version`.  Ideally there would be a way to either explicitly control the MSRV, or a way to pick the *highest* MSRV in the workspace.

**Comment by Stebalien** [comment]
I've just run into this issue in the tempfile crate. I support older compiler versions, but that doesn't mean I want to test/build with outdated dependencies when using a recent compiler.  I've worked around this in CI by [replacing](https://github.com/Stebalien/tempfile/pull/327/changes/b690bc47c054b780addc4c6d2d74e6784304758b) `rust-version` with the version of the current compiler, but this is an ugly hack that only works in CI.

TBH, I'm really confused at why/how this got implemented this way:

1. The rust `rust-version` field specifies the minimum rust version my crate supports so that my _dependent_ crates can pick a version of my crate supported by their compiler. This field *should not* affect how cargo selects dependencies for *my* `Cargo.lock` file (which my dependencies don't care about).
2. In terms of locking the compiler version for _my_ crate, isn't that what `rust-toolchain.toml` was for?

Cargo probably should use the `rust-version` field to validate that the version constraints in my `Cargo.toml` file can be satisfied by my declared MSRV. The important distinction is, again, things that _my_ crate cares about versus things my _dependent_ crates care about.

**Comment by epage** [comment]
> TBH, I'm really confused at why/how this got implemented this way:

https://rust-lang.github.io/rfcs/3537-msrv-resolver.html does a thorough analysis of different workflows against different solutions.

**Comment by Stebalien** [comment]
I now have a better understanding of why it's done this this way, but it still leaves me hanging as a library maintainer. Specifically:

> ❌ For library and tool maintainers, they are absorbing costs from these less common use cases
They could shift these costs to those that need old versions by switching to the “Latest MSRV” workflow by allowing their users to backport fixes to prior MSRV releases

For applications/tools, this makes sense. Declare a recent MSRV and let users stuck on old compilers deal with backporting, downgrading dependencies, etc.

For (transitive) dependencies/libraries, it's not so easy to simply "backport fixes". There's a reason much of the ecosystem still declares 1.85 as the MSRV. It looks like the RFC at one point considered a `CARGO_RESOLVER_PRECEDENCE=rustc` escape hatch, which is exactly what I need.

**Comment by ijackson** [comment]
> For (transitive) dependencies/libraries, it's not so easy to simply "backport fixes".

Quite so.  As a downstream I very much value libraries that maintain a wide MSRV range and as a library maintainer I try to do the same for my downstreams.

> There's a reason much of the ecosystem still declares 1.85 as the MSRV

1.85!  serde is at 1.56.

The most popular library by downloads that I maintain is shellexpand and there I have no plans to increase the MSRV from 1.31.

cargo ought to be supporting our work and not trying to force particular choices on us. As library maintainers we are in the best position to make our own tradeoffs.

**Comment by matklad** [comment]
@Stebalien for CI, you might use `cargo generate-lockfile` when testing MSRV, and `cargo generate-lockfile --ignore-rust-version` when testing stable/beta/nightly.

**Comment by ijackson** [comment]
> [@Stebalien](https://github.com/Stebalien) for CI, you might use `cargo generate-lockfile` when testing MSRV, and `cargo generate-lockfile --ignore-rust-version` when testing stable/beta/nightly.

Right.  In one of my projects, I do roughly that.

For situations where the library author needs more precise control, I would like an explicit command line option to specify the Rust version, as I requested in #16672.  AIUI this is in some sense on the roadmap?

**Comment by Stebalien** [comment]
> @Stebalien for CI, you might use cargo generate-lockfile when testing MSRV, and cargo generate-lockfile --ignore-rust-version when testing stable/beta/nightly.

That sort-of, kind-of works; but it makes it harder to reproduce CI failures locally. My new solution is to:

1. Set `resolver.incompatible-rust-versions` to allow in my repo's `.cargo/config.toml` file. That way my local dev environment is as close to CI as I can get it (so I don't pull my hair out trying to reproduce CI failures).
2. Set `CARGO_RESOLVER_INCOMPATIBLE_RUST_VERSIONS=fallback` in CI when testing MSRV.

While this "works for me", the solution is non-trivial and failure to configure one's project this way will lead to bugs that only show up in downstream dependent packages due to them having more recent dependencies (by selecting a more recent MSRV). I expect this will become a common stumbling block.
